### ISSUE TITLE
Loading and Saving data

### ISSUE THREAD

**Issue by abauer** [JPII/NavalBattle#67]
This is a combination of  multiple issues including
#58, SettingsAttributes fails to load,
#46, Implement LoginOptionsWindow,
#57, Ability to save data,
#47, LoginWindow should save last good username,
#49, The ability to save chunks and the current state of the grid,

This issue includes (based off of aforementioned issues:
1) a working read/write file system and location:
2) a interface with the LoginOptionsWindow
3) an "OOP-based system of accessing stored data, preferably in separate files as needed."
4) username saving
5) a system of saving world data
6) saving of world data

**Comment by maximusvladimir** [comment]
Wo wo wo bro, this isn't suppose to be this way... The settings are going to be independent of the worlds themselves. issue #57 and issue #58 shouldn't be included in this.

**Comment by abauer** [comment]
actually all the data needs to be stored somewhere; yes they will be saved in different places but the reader / writer encompasses all the data, thus it does include #57 and #58 : But you are correct in that #57 and #58 are separate from the Options

**Comment by ThomasGaubert** [comment]
I mentioned an idea about how the file saving should work. Hopefully this makes sense:
- Use by creating an object from `PreferenceManager.java`
  - Ideally one object is for NavalBattle global preferences and can be accessed by `NavalBattle.getPreferenceManager()`
  - Each instance of the PreferenceManager handles one individual file, stored somewhere safe (we should look at Hawkridge, which uses the Minecraft launcher's not so open source code)
  - New instances can be made when data needs to be stored separately (for example with the render engine)
- All data objects are stored with a key and a value. Keys are used to access data and the value is obviously what is being stored.
- This works similar to that of a `HashMap`
- This format of saving data would be compatible with RoketGamer's data API as it will most likely work the same
- `PreferenceManager` can be extended to support JSON and other data formats
  - Data format for plaintext needs to be decided assuming it will use plaintext

Class names are obviously subject to change as we see fit.

**Comment by maximusvladimir** [comment]
Thanks to the new renderer being implemented, chunk data should be handled more effectively.

For example the file "3b.chk", "zZ.chk", "V8.chk", would correspond to the chunk at coordinates (3,11), (35,61), (56,8). (Kinda how Minec\* handles it).

Therefore, the maximum world size, can currently be 61x61, until I implement super-chunks (which could yield a maximum world size of 13845841x13845841.) (Yes, world size is now dynamic).

Now you are probably asking: Well the game was REALLY slow when I ran it in a really big sized window, and it is already kinda slow, how do you expect the game to be able to handle a world the size of 13845841x13845841?

Well the answer is actually simple (not really). What happens is, each super-chunk would be stored either in memory, or on the disk, to help with memory usage. Each entity on the chunk would have a shared image, through a global image manager. When the chunk is first generated, a image will be created with the terrain data. The terrain data will be reserved for future uses (like the OmniMap, etc). When a chunk is updated, it will call each entity's updator. The World Day/Night cycle will be called on a global level (through World.java), so there is very little load on each Chunk. The chunks are being stored in a single dimensional array, to help with performance. The layer rendering order for the chunks will work the following way: Terrain, Day/Night overlay, Entities, Screen shadower, OmniMap, and finally the GUI. You may be asking the following: Why isn't the Day/Night overlay applied to the Entities??? Well, because I believe that the ships, whales, etc, should handle day and night on their own (so they could have lights on them during the night. Anyways back to the loading and saving stuff, basically, I think that this new method that I have thought up, is going to let the game run really smoothly, and save and load chunks really easy.

Also, since each chunk is 100x100 pixels, and an entity is 50 pixels, we can store 4 entities, maximum.

Each chunk file would have a structure of the following:

``` javascript
# chunk d4 - DO NOT MODIFY WITHOUT FIRST UNDERSTANDING WHAT YOU ARE DOING
seed: 3491203190402
terrain: 5,109,89,230,191,39,19,10 .......
width: 100
height: 100
entity00: id=2,health=4,anger=20
entity01: id=-1
entity10: id=-1
entity11: id=10,health=100,...
```

So this kind of cover topic no. 3,5, and 6.

**Comment by ThomasGaubert** [comment]
So essentially each chunk is a 2x2 grid in the game?

If I'm understanding everything correctly the game currently has each tile (grid space) as a entity with the image data, etc. (this uses a lot of memory). Will this be the case again or will chunks and the map data be stored on a separate layer?

**Comment by maximusvladimir** [comment]
Chunks and map data are going to be _rendered_ in different layes, but they are going to pragmatically be together.
Lets see if I can make a graphic to help:

![Graphic showing stuff about the organization](https://dl.dropbox.com/u/11164236/navalbattle/organizationmap.png)

**Comment by maximusvladimir** [comment]
Sorry, in the graphic above, I swapped the Entity 10 and the Entity 01.

**Comment by ThomasGaubert** [comment]
Wow that's cool. Thanks!

**Comment by maximusvladimir** [comment]
This can't really be done until I have finished Pavo.

**Comment by maximusvladimir** [comment]
I propose an idea: we have a class or abstract method that would handle saving appropriatly.

**Comment by abauer** [comment]
ok - i guess that works

**Comment by abauer** [comment]
1) location has been added as of FileUtils.java located in jpii.com.navalbattle.util

**Comment by maximusvladimir** [comment]
Is this still being worked on?

**Comment by maximusvladimir** [comment]
We should be able to work on it now.

**Comment by ThomasGaubert** [comment]
I'll handle LoginOptions (#121 and others...)

**Comment by maximusvladimir** [comment]
If you want to see how we are loading settings, take a look at MaxTests.java (DO NOT MODIFY THE FILE THOUGH! IT HAS MY NAME ON IT!!!!). The sample settings file is in %APPDATA%/.navalbattle/settings.ini.

**Comment by ThomasGaubert** [comment]
Ok, thanks,

**Comment by maximusvladimir** [comment]
As of commit id bff1dc0ace475e04962f313949dbc790e1328979, SettingsWriter has been implemented.

**Comment by abauer** [comment]
so progress on this issue? someone want to make a checklist for this?

**Comment by ThomasGaubert** [comment]
Yeah we need to make a list of what we need to save at any point in time.

I'll start:
- LoginOptions
- Saved games
- RoketGamer cache files (I may handle this on my own in the RoketGamer package)

**Comment by maximusvladimir** [comment]
LoginOptions -> SettingsReader/SettingsWriter

Game Settings (user picked settings that are not default) -> SettingsReader/SettingsWriter (same file as LoginOptions)

Game States (last game window position, last window size, last internal game windows configuration) -> (same file as LoginOptions)

**Comment by abauer** [comment]
we don't need to save window information: as most is either hardcoded or non adjustable...

**Comment by maximusvladimir** [comment]
Ok thats find, Also my thoughts is that everything except world saves should be placed in one centralized SettingsReader/Writer file.

**Comment by abauer** [comment]
ok thats a good idea, kinda like a utility

**Comment by ThomasGaubert** [comment]
Exactly. Possibly using JSON, YAML, XML, or a custom markup language.

**Comment by maximusvladimir** [comment]
?

.navalbattle
|
--- settings.ini (LoginOptions, GameSettings, etc)
|
--- saves
....|
....--- MyGame1.gmf
....|
....--- MyGame2.gmf
...|

**Comment by ThomasGaubert** [comment]
That would work.

**Comment by maximusvladimir** [comment]
And we would use SettingsReader/SettingsWriter? (Look in MaxTests.java for examples on how to use).

(The game saves are using it)

**Comment by ThomasGaubert** [comment]
Yes, that's the plan I believe.

**Comment by maximusvladimir** [comment]
Ok.

**Comment by abauer** [comment]
so whats next on this issue?

**Comment by ThomasGaubert** [comment]
Implementing it afaik?

**Comment by ThomasGaubert** [comment]
This is two days overdue. This needs to be fixed ASAP.

For everyday we are behind, it's less time to work on future milestones. Due dates won't change.

**Comment by maximusvladimir** [comment]
I will work on this today.

**Comment by ThomasGaubert** [comment]
We have a system in place for this.

**Comment by maximusvladimir** [comment]
Did we ever finish this? I can't seem to save my game.

**Comment by ThomasGaubert** [comment]
Not sure, I thought we had it implemented.

**Comment by abauer** [comment]
the system was written, but i don't know if you ever finished the saving part

**Comment by maximusvladimir** [comment]
I did write the saving part, but due to the fact that we are using turns, entities, etc, I don't think it will save anymore.

Someone should test out the `save` command, and see what it does.

@abauer We wrote the saving part, but not a loading part (there's a structure for it). I don't think the saving part works anymore though.

**Comment by ThomasGaubert** [comment]
Ah okay, it makes sense now.

**Comment by ThomasGaubert** [comment]
Reopened for now.

**Comment by ThomasGaubert** [comment]
I don't know why this is assigned to me...

**Comment by maximusvladimir** [comment]
That's a good question.

**Comment by ThomasGaubert** [comment]
At this point do we want saving? It could add to the difficulty to have to play it in a sitting. Not to mention we need to come up with a file format (that accounts for features that haven't been added).

It can always be added later.

**Comment by maximusvladimir** [comment]
I can't understand why this was marked for this issue, and we still don't have a working game....

**Comment by ThomasGaubert** [comment]
Exactly, we can't save stuff if we don't know what to save.

**Comment by ThomasGaubert** [comment]
Removed milestone.

**Comment by abauer** [comment]
agree with these recent statements

**Comment by maximusvladimir** [comment]
So saving data works.

Now all that is left is to load it.

**Comment by maximusvladimir** [comment]
Things that should be saveable:
- [x] Seed (so basically chunks)
- [x] Entity data
- [x] Was the game registered online (in RoketGamer)?
- [x] The name of the game/world.
- [x] The version of the game.
- [ ] Time
- [ ] Current turn and other turn stuff (I still have no clue how the framework for this stuff works).
- [ ] ???

**Comment by abauer** [comment]
wat about turn? um you should save the turn number and player information (@darkwarhero) for AI

**Comment by maximusvladimir** [comment]
Or, you could just get the seed for the AI, and then do something like this:

``` Java
// Original Game (the one that is being saved.)
Rand aiRandom;

public void onSave() {
saveAttribute("numberofrandomnumbersgenerated", aiRandom.getNumCalculated());
saveAttribute("aiseed", aiRandom.getSeed());
}

// The "New" Game (the one that is being loaded.)

Rand aiRandom;

public void onLoad() {
aiRandom = new Random(loadAttributeAsLong("aiseed"));
aiRandom.restore(loadAttributeAsLong("numberofrandomnumbersgenerated"));
}

```

**Comment by ThomasGaubert** [comment]
Do we really want/need this?

**Comment by maximusvladimir** [comment]
Idk.

**Comment by ThomasGaubert** [comment]
Moved it off the milestone for now since it isn't critical. I'd like to get this game shipped :shipit:
